### PR TITLE
Restructure Unix interop code

### PR DESCRIPTION
**Pull request by stephentoub**
As we've discussed offline, we want to change how we're structuring our interop code, in particular to make it easier to share interop definitions across assemblies.  This commit is an initial restructuring of our existing Unix interop code to follow the new approach.

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @weshaggard, @ellismg

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by ellismg** [comment]
Looks good.  In my review, I paid careful attention to the modified files and only glanced at the new ones, assuming that it was more code motion than code authoring.

**Comment by stephentoub** [comment]
Thanks, Matt.  Good suggestion about the wiki.  I'm going to merge these changes to unblock some other things, and then separately I'll add info to the wiki on the approach.

File: src/Common/src/Interop/Unix/libc/Interop.Permissions.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+internal static partial class Interop
+{
+    internal static partial class libc
+    {
+        [Flags]
+        internal enum Permissions
```
Comment (weshaggard) [inline src/Common/src/Interop/Unix/libc/Interop.Permissions.cs]: Was the guidance for these enums to be under lib? Or directly in Interop?

File: src/Common/src/Interop/Unix/libc/Interop.getegid.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class libc
+    {
+        [DllImport(Libraries.Libc)]
+        internal static extern int getegid();
+    }
+}
```
Comment (weshaggard) [inline src/Common/src/Interop/Unix/libc/Interop.getegid.cs]: Perhaps just a github thing but there seems to be an inconsistency with the newlines at the end of the files.

File: src/Microsoft.Win32.Primitives/src/Microsoft.Win32.Primitives.csproj [hunk 3]
```diff
@@ -27,9 +27,9 @@
     <Compile Include="System\ComponentModel\Win32Exception.Windows.cs" />
   </ItemGroup>
   <ItemGroup Condition=" '$(OS)' == 'Unix' ">
-    <Compile Include="$(CommonPath)\Interop\Interop.strerror.Unix.cs" />
-    <Compile Include="Interop\Interop.Unix.cs" />
     <Compile Include="System\ComponentModel\Win32Exception.Unix.cs" />
+    <Compile Include="$(CommonPath)\Interop\Unix\Interop.Libraries.cs"><Link>Interop\Interop.Libraries.cs</Link></Compile>
+    <Compile Include="$(CommonPath)\Interop\Unix\libc\Interop.strerror.cs"><Link>Interop\Interop.strerror.cs</Link></Compile>
```
Comment (weshaggard) [inline src/Microsoft.Win32.Primitives/src/Microsoft.Win32.Primitives.csproj]: For the the Link I've been suggesting that people use Common in the front so it easy to see these come from common. So something like Common\Interop\Interop.Libaries.cs

File: src/System.Console/src/System.Console.csproj [hunk 4]
```diff
@@ -53,22 +53,19 @@
...
-    </Compile>
-    <Compile Include="$(CommonPath)\Interop\Interop.strerror.Unix.cs">
-      <Link>Common\Interop\Interop.strerror.Unix.cs</Link>
-    </Compile>
-    <Compile Include="$(CommonPath)\Microsoft\Win32\SafeHandles\SafeFileHandle.Unix.cs">
-      <Link>Common\Microsoft\Win32\SafeHandles\SafeFileHandle.Unix.cs</Link>
-    </Compile>
+    <Compile Include="$(CommonPath)\Microsoft\Win32\SafeHandles\SafeFileHandle.Unix.cs"><Link>Common\Microsoft\Win32\SafeHandles\SafeFileHandle.Unix.cs</Link></Compile>
+    <Compile Include="$(CommonPath)\System\IO\StringBuilderCache.cs"><Link>Common\System\IO\StringBuilderCache.cs</Link></Compile>
+
```
Comment (weshaggard) [inline src/System.Console/src/System.Console.csproj]: I'm not sure I think the Link on the same line is better. Does VS reformat these when you save the project there? if so I would just put them on their own lines.

**Comment by weshaggard** [comment]
@stephentoub sorry for being late to the review. I've added a couple comments in the commits but nothing major.

File: src/Common/src/Interop/Unix/libc/Interop.Permissions.cs (hunk 1 above)
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libc/Interop.Permissions.cs]: I think they should be in the library type if they're defined by header files from that library, which they are in this case.  I'll make sure the guidance is clear on that (unless you disagree).

**Comment by stephentoub** [comment]
Thanks, @weshaggard, I'll submit a another PR to address the line endings, reformat the .csproj files to use multiple lines for links, and to use "Common" in the links.

**Comment by stephentoub** [comment]
@ellismg, FYI, I've added a wiki page with the interop guidelines at https://github.com/dotnet/corefx/wiki/Structuring-Interop-Code

**Comment by weshaggard** [comment]
Thanks @stephentoub.

**Pull request merged**